src/crab/type_domain.cpp

```diff
@@ -0,0 +1,714 @@
+    return !(p1 == p2);
+}
+
+std::ostream& operator<<(std::ostream& o, const ptr_with_off_t& p) {
+    o << get_reg_ptr(p.r) << "<" << p.offset << ">";
+    return o;
+}
+
+bool operator==(const ptr_no_off_t& p1, const ptr_no_off_t& p2) {
+    return (p1.r == p2.r);
+}
+
+bool operator!=(const ptr_no_off_t& p1, const ptr_no_off_t& p2) {
+    return !(p1 == p2);
+}
+
+std::ostream& operator<<(std::ostream& o, const ptr_no_off_t& p) {
+    return o << get_reg_ptr(p.r);
+}
+
+std::ostream& operator<<(std::ostream& o, const reg_with_loc_t& reg) {
+    o << "r" << static_cast<unsigned int>(reg.r) << "@" << reg.loc->second << " in " << reg.loc->first << " ";
+    return o;
+}
+
+bool reg_with_loc_t::operator==(const reg_with_loc_t& other) const {
+    return (r == other.r && loc == other.loc);
+}
+
+std::size_t reg_with_loc_t::hash() const {
+    // Similar to boost::hash_combine
+    using std::hash;
+
+    std::size_t seed = hash<register_t>()(r);
+    seed ^= hash<int>()(loc->first.from) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+    seed ^= hash<int>()(loc->first.to) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+    seed ^= hash<int>()(loc->second) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+
+    return seed;
+}
+
+std::ostream& operator<<(std::ostream& o, const stack_t& st) {
+    o << "Stack: ";
+    if (st.is_bottom())
+        o << "_|_\n";
+    else {
+        o << "{";
+        for (auto s : st.m_ptrs) {
+            o << s.first << ": ";
+            print_ptr_type(s.second);
+            o << ", ";
+        }
+        o << "}";
+    }
+    return o;
+}
+
+std::ostream& operator<<(std::ostream& o, const ctx_t& _ctx) {
+
+    o << "type of context: " << (_ctx.m_packet_ptrs.empty() ? "_|_" : "") << "\n";
+    for (const auto& it : _ctx.m_packet_ptrs) {
+        o << "  stores at " << it.first << ": " << it.second << "\n";
+    }
+    return o;
+}
+
+ctx_t::ctx_t(const ebpf_context_descriptor_t* desc)
+{
+    if (desc->data != -1)
+        m_packet_ptrs[desc->data] = crab::ptr_no_off_t(crab::region::T_PACKET);
+    if (desc->end != -1)
+        m_packet_ptrs[desc->end] = crab::ptr_no_off_t(crab::region::T_PACKET);
+}
+
+std::optional<ptr_no_off_t> ctx_t::find(int key) const {
+    auto it = m_packet_ptrs.find(key);
+    if (it == m_packet_ptrs.end()) return {};
+    return it->second;
+}
+
+
+std::ostream& operator<<(std::ostream& o, const register_types_t& typ) {
+    if (typ.is_bottom())
+        o << "_|_\n";
+    else {
+        for (const auto& v : *(typ.m_all_types)) {
+            o << v.first << ": ";
+            print_ptr_type(v.second);
+            o << "\n";
+        }
+    }
+    return o;
+}
+
+register_types_t register_types_t::operator|(const register_types_t& other) const {
+    if (is_bottom() || other.is_top()) {
+        return other;
+    } else if (other.is_bottom() || is_top()) {
+        return *this;
+    }
+    live_registers_t out_vars;
+    for (size_t i = 0; i < m_vars.size(); i++) {
```

Comment:
I think this join is wrong. Let' say we have this code
```
if (*) {
  loc_i: R1 has some type T;
} else {
  loc_j: R1 has some type T;
}
```
This join will lose the type of R1 because  it will not find `loc_i` in `other`.
So you need to iterate directly over`m_all_types`

Reply:
what I do is, look for the type of `R1` in if-branch and else-branch. Then check if they are the same. I am not sure why location (loc_i or loc_j) is important. If we do not find loc_i in other, we should be fine. We look for registers values, not specific locations